src/reader/async_tokio.rs

```diff
@@ -0,0 +1,462 @@
+pub struct TokioAdapter<R>(pub R);
+
+impl<R: AsyncBufRead + Unpin> TokioAdapter<R> {
+    impl_buffered_source!('b, 0, async, await);
+}
+
+impl<R> Deref for TokioAdapter<R> {
+    type Target = R;
+
+    fn deref(&self) -> &R {
+        &self.0
+    }
+}
+
+impl<R> DerefMut for TokioAdapter<R> {
+    fn deref_mut(&mut self) -> &mut R {
+        &mut self.0
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+impl<R: AsyncBufRead + Unpin> Reader<TokioAdapter<R>> {
+    /// Creates a `Reader` that reads from a given reader asynchronously.
+    pub fn from_async_reader(reader: R) -> Self {
+        Self::from_reader(TokioAdapter(reader))
+    }
+
+    /// An asynchronous version of [`read_event_into()`]. Reads the next event into
+    /// given buffer.
+    ///
+    /// > This function should be defined as
+    /// > ```ignore
+    /// > pub async fn read_event_into_async<'b>(
+    /// >     &mut self,
+    /// >     buf: &'b mut Vec<u8>
+    /// > ) -> Result<Event<'b>>;
+    /// > ```
+    /// > but Rust does not allow to write that for recursive asynchronous function
+    ///
+    /// This is the main entry point for reading XML `Event`s when using an async reader.
+    ///
+    /// See the documentation of [`read_event_into()`] for more information.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # tokio_test::block_on(async {
+    /// # use pretty_assertions::assert_eq;
+    /// use quick_xml::events::Event;
+    /// use quick_xml::reader::Reader;
+    ///
+    /// // This explicitly uses `from_async_reader("...".as_bytes())` to use a buffered
+    /// // reader instead of relying on the zero-copy optimizations for reading
+    /// // from byte slices, which is provides the sync interface anyway.
+    /// let mut reader = Reader::from_async_reader(r#"
+    ///     <tag1 att1 = "test">
+    ///        <tag2><!--Test comment-->Test</tag2>
+    ///        <tag2>Test 2</tag2>
+    ///     </tag1>
+    /// "#.as_bytes());
+    /// reader.trim_text(true);
+    ///
+    /// let mut count = 0;
+    /// let mut buf = Vec::new();
+    /// let mut txt = Vec::new();
+    /// loop {
+    ///     match reader.read_event_into_async(&mut buf).await {
+    ///         Ok(Event::Start(_)) => count += 1,
+    ///         Ok(Event::Text(e)) => txt.push(e.unescape().unwrap().into_owned()),
+    ///         Err(e) => panic!("Error at position {}: {:?}", reader.buffer_position(), e),
+    ///         Ok(Event::Eof) => break,
+    ///         _ => (),
+    ///     }
+    ///     buf.clear();
+    /// }
+    /// assert_eq!(count, 3);
+    /// assert_eq!(txt, vec!["Test".to_string(), "Test 2".to_string()]);
+    /// # }) // tokio_test::block_on
+    /// ```
+    ///
+    /// [`read_event_into()`]: Reader::read_event_into
+    pub fn read_event_into_async<'r, 'b: 'r>(
+        &'r mut self,
+        buf: &'b mut Vec<u8>,
+    ) -> Pin<Box<dyn Future<Output = Result<Event<'b>>> + 'r>> {
```

Comment:
The `'_self` name actually does not clarify nothing. `'r` is abbreviation of `'reader` which seems too long for me and I don't see good abbreviation. So this method should read as: "returns `Event` which borrow data from `buf` and that buffer live longer than reader _self_"

Reply:
`'_self` was just an example because it's a name I like to use in my projects. Maybe we could use `'rdr` here for the reader and `'buf` for the buffer to have more descriptive names for lifetimes than `'r` and `'b`.

Reply:
In the end I've changed `'r` to `'reader`, but leave `'b` -- I think it is quite descriptive and is already used in many places to indicate the lifetime of the buffer. In addition, with this choice of names, `mut` are beautifully aligned in the code and documentation :innocent:

Reply:
Okay, sounds good :)